### DISCUSSION TITLE
Cypress component test throws "process is not defined"

### DISCUSSION THREAD

**Discussion by tomascaroponce** [cypress-io/cypress#26645] [Questions and Help]
I'm trying to run component tests on a PWA made with Vue 2 and Webpack. When I try to run them I get an error saying `process is not defined` and this prevents me from doing the tests I want.
![image](https://user-images.githubusercontent.com/132400728/235780546-8511d2cb-53d6-4417-967b-98a0bfe67854.png)
The dependencies that my project has are the following:
```
"dependencies": {
    "compressorjs": "^1.1.1",
    "dotenv": "^16.0.3",
    "jquery": "^3.4.1",
    "lodash": "^4.17.21",
    "npm": "^6.10.0",
    "register-service-worker": "^1.5.2",
    "vue": "^2.6.14",
    "vue-router": "^3.0.1",
    "vuex": "^3.0.1",
    "vuex-persist": "^3.1.3"
  },
  "devDependencies": {
    "@babel/cli": "^7.2.3",
    "@babel/core": "^7.12.16",
    "@babel/polyfill": "^7.2.5",
    "@babel/preset-env": "^7.21.4",
    "@vue/cli-plugin-babel": "~5.0.0",
    "@vue/cli-plugin-eslint": "~5.0.0",
    "@vue/cli-plugin-pwa": "^3.1.1",
    "@vue/cli-service": "~5.0.0",
    "babel-helper-vue-jsx-merge-props": "^2.0.3",
    "babel-plugin-syntax-jsx": "^6.18.0",
    "babel-plugin-transform-vue-jsx": "^3.7.0",
    "babel-preset-env": "^1.7.0",
    "copyfiles": "^1.2.0",
    "css-loader": "^6.7.3",
    "cypress": "^12.11.0",
    "node-sass": "^4.12.0",
    "sass-loader": "^7.1.0",
    "vue-cli-plugin-vuetify": "^0.1.6",
    "vue-loader": "^15.10.1",
    "vue-template-compiler": "^2.7.14",
    "web-push": "^3.3.3",
    "webpack": "^5.80.0",
    "workbox-cli": "^3.5.0"
  }
```

My webpack config file looks like this:
```
const { VueLoaderPlugin } = require('vue-loader')

module.exports = {
  entry: ["@babel/polyfill", "./app.vue"],
  module: {
    rules: [
      {
        test: /\.css$/,
        use: [
          'css-loader'
        ]
      },
      {
        test: /\.vue$/,
        loader: 'vue-loader'
      }
    ]
  },
  plugins: [
    new VueLoaderPlugin()
  ]
};
```

My cypress config file looks like this:

```
const { defineConfig } = require("cypress");
const webpackConfig = require('./src/webpack.config');

module.exports = defineConfig({
  component: {
    devServer: {
      framework: "vue",
      bundler: "webpack",
      webpackConfig
    },
  },
});
```

And my test file looks like this:

```
import Loading from './Loading.vue'

describe('<Loading />', () => {
  it('renders', () => {
    cy.mount(Loading)
  })
})
```

I suppose it is a conflict between dependencies or a problem with some loader library.

I tried to implement all the solutions proposed in forums but none of them worked for me. I also created a new project with vue 2 and the tests ran perfectly, but I couldn't find what is the reason for the problem.

**Comment by portyu9** [comment]
Hi @tomascaroponce 

The screenshot shows that this is coming from your app code, specifically `src/plugins/firebase.js`:

```js
apiKey: process.env.VUE_APP_FIREBASE_API_KEY
```

The important difference between your working Vue CLI app and the Cypress Component Testing setup is the webpack config.

Vue CLI normally replaces `process.env.VUE_APP_*` references **at build time** using webpack's `DefinePlugin`. Your Cypress config replaces the Vue CLI webpack configuration with this custom config:

```js
const webpackConfig = require('./src/webpack.config')
```

That config only contains `vue-loader`/`css-loader`, so the Vue CLI environment-variable injection is no longer present. With webpack 5 there is also no automatic browser polyfill for `process`, hence the runtime `ReferenceError`.

For Cypress 12.x + Vue 2, I would avoid recreating the Vue CLI webpack config and instead reuse the resolved Vue CLI configuration, for example:

```js
const { defineConfig } = require('cypress')
const webpackConfig = require('@vue/cli-service/webpack.config.js')

module.exports = defineConfig({
  component: {
    devServer: {
      framework: 'vue',
      bundler: 'webpack',
      webpackConfig,
    },
  },
})
```

Alternatively, if you must keep your custom webpack config, add `DefinePlugin` entries for the `VUE_APP_*` variables you actually expose to the client. Simply adding a `process` polyfill is not the ideal fix here because these values are supposed to be compile-time substitutions in a Vue CLI app.

There is also a separate dependency mismatch in the posted `package.json`:

```text
vue:                  2.6.14
vue-template-compiler: 2.7.14
```

Those should be matching Vue 2 versions.

One current-version note: Cypress 16's supported Vue Component Testing integration is now Vue 3+. So for an application that must remain on Vue 2, this is a legacy Cypress CT setup; otherwise migrating the app to Vue 3 is the current supported path.

Please mark this answer as accepted if it helped, thank you.
